### ISSUE TITLE
Support Custom Pod Status for PodReadinessGate to Block Premature Pod Termination

### ISSUE THREAD

**Issue by BrianChristie** [kubernetes-sigs/aws-alb-ingress-controller#905]
Presently when using IP target mode on the ALB, it's difficult to guarantee zero downtime rolling updates. This is because there is additional state in the ALB Target Group which is not propagated back to Kubernetes: `initial`,`draining`, and `healthy`. 

This was discussed in:  https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/660 & https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/814

This issue is to discuss adding support to the ALB controller for setting a custom Pod Status which mirrors the ALB TargetGroup. This can then be used with `PodReadinessGate` and `MaxUnavailable` / `PodDisruptionBudget` to ensure a rolling update does not proceed faster than the new Pods becoming healthy on the ALB.

- KEP: kubernetes/enhancements#580
- Docs: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate

**Comment by devkid** [comment]
I'm working on this.

**Comment by devkid** [comment]
For reconciling, the controller would need to query every pod behind every service that is a backend for every ingress that is managed by the alb ingress controller, on every iteration – this would potentially introduce more load on the API server. Would it make sense to add an annotation on the ingress (or service), like:
* `alb.ingress.kubernetes.io/enable-pod-condition` (or something other wording?)

(and possibly a CLI argument to set a default value for this option?) so that users can decide for their own if they want to enable this?

**Comment by devkid** [comment]
I won't have time to continue working on this in the next two weeks, but I have pushed an initial implementation here: https://github.com/devkid/aws-alb-ingress-controller/tree/feature/pod-readiness-gate (it compiles and has documentation, but I didn't test it yet nor did I adjust the unit tests).

It works by reconciling the pod conditions in the same step where it reconciles the ALB target group state. I think it wouldn't work in its current state because right now only the `Ready` endpoints of a service are registered with the ALB. When a pod declares the ALB readiness gate, it would never turn `Ready` because it would wait for the readiness gate (turn healthy in ALB), which would never be fulfilled because the pod was never registered with the ALB (because it was not `Ready` yet) - so endless loop. To solve this, the ALB ingress controller would need to support the `publishNotReadyAddresses` attribute of a Kubernetes services and it would need to be set to `true` if one would want to use this pod readiness gate feature (then all pods - also unready ones - would be registered with the target group, pass the health check and only then appear Ready in kubernetes).

If someone feels like picking up from where I left, feel free to do so. Otherwise I would probably continue in ~2-3 weeks.

**Comment by tecnobrat** [comment]
Hey @devkid -- just want to checkin to see if you have had a chance to get back to this change at all.

This is impacting us and we're open to trying to help with any PRs that need to be done if you need help.

Thanks!

**Comment by boynux** [comment]
Hi,
We hit this issue in our new EKS cluster which prevents us moving forward with the production workload, any progress with the proposed PR?

**Comment by devkid** [comment]
Well, the functionality is there, I'm still waiting for reviews. You can try the code from my branch if you want to, that would be helpful feedback already.
I'll be meeting some EKS people tomorrow and will try to push them to get this reviewed by some of them.

**Comment by devkid** [comment]
Btw a workaround would be to have at least 2 replicas and `maxUnavailable` = 0 and an `initialDelaySeconds` that is large enough to cover the time it takes to register a target with an target group and for it to become healthy; this should prevent the last pod from terminating until this delay elapses for the second new pod (which is not yet registered with the target group, but in the meantime the first new pod was ready long enough to be healthy in the target group).

**Comment by gzzo** [comment]
@devkid Would a pod be registered on the ALB before it's ready?

I had a similar hypothesis a while ago and I tried with 6 replicas, 0 `maxUnavailable`, and 60 `minReadySeconds` but I still saw 504s with IP target mode.

**Comment by nirnanaaa** [comment]
@devkid there are still some issues with your approach. I'm trying to come up with a follow-up PR to yours so we can maybe share the work on this.

**Comment by wirescrossed** [comment]
We are having a similar issue as well. When pods are terminated the ALB Ingress Controller rapidly updates the ALB target group however the changes take a few seconds to take effect which causes a short period where clients may get 504s.

We are using IP mode meaning each Pod is in the target group since we require the real ALB source IP not the source IP kube-proxy uses in order to use NetworkPolicies.

To workaround this we are taking the approach of running kube-proxy without source NAT (https://kubernetes.io/docs/tutorials/services/source-ip/) though this has draw backs as well but masks the issue to a degree.

What I would really like to see is one of two things:
- ALB to keep pace with the rapid changes in k8s
- k8s give the ALB time to catch up by delaying termination of the pod.

**Comment by nirnanaaa** [comment]
I would be willing to continue @devkid s and my work on this PR. Maybe in collaboration with someone from AWS that could give us insight on the Polling-thing i was talking about earlier.

/cc @andrewtandoc

**Comment by iverberk** [comment]
We are hitting this as well and it's a blocker for production workloads. @devkid @nirnanaaa I was going to review and finish this PR also, so let's explore the problem space and collaborate on a design that we can implement. It is actually very similar to what Google needed to solve for their loadbalancers (23:00 in this video https://youtu.be/Vw9GmSeomFg)

**Comment by nirnanaaa** [comment]
@iverberk you're absolutely right. We were able to solve most of the challenges by taking a look at ingress-gce (basically this piece is the relevant one: https://github.com/kubernetes/ingress-gce/tree/5b2fd1df8634ca02aaa10072d7769a2544d39cb5/pkg/neg/readiness)

Except for the poller everything works just fine (didn't figure out a way, to make that work).

Do you suggest starting of fresh or should we base the work on our previous efforts? Maybe you have something to overcome that issue, already? See @devkid s PR and my changes here: `https://github.com/kubernetes-sigs/aws-alb-ingress-controller/compare/master...nirnanaaa:feature/pod-readiness-gate#diff-c6555fa828feb3fa0daf4ff60736047dR130`

**Comment by iverberk** [comment]
Not necessarily starting of fresh, it seems the implementation is almost there, just didn't want to duplicate efforts. I didn't know about that ingress-gce code, but will check it out.

I guess you suggest polling to detect changes to target groups (targets becoming unhealthy etc.) that happen outside the aws-alb-ingress-controller purview and need reconciling with Pod readiness gates?

**Comment by devkid** [comment]
@nirnanaaa @iverberk do we maybe want to meet to discuss the next steps? I would send out an invite tomorrow from my company email.

**Comment by nirnanaaa** [comment]
@devkid 

@iverberk and I were already in talks on slack, I could not find you on the k8s slack?

**Comment by wirescrossed** [comment]
We were able to slow down the termination of pods to let the ALB catch up thus preventing 504s by using lifecycle prestop hook

```
        lifecycle:
          preStop:
            exec:
              command:
              - /bin/bash
              - -c
              - sleep 25
              - apachectl -k graceful-stop
```

**Comment by tecnobrat** [comment]
@nirnanaaa try maybe reaching out to @M00nF1sh?  They are on the k8s slack as well.

**Comment by nirnanaaa** [comment]
will do, thanks!

**Comment by iverberk** [comment]
@nirnanaaa @devkid were you able to make some progress on this? I think we need to pick this up as the ingress controller is unable to do zero-downtime deployments in the current state.

**Comment by nirnanaaa** [comment]
no progress, yet. Also was gone on vacation and had some internal work to do. I was talking to @M00nF1sh to find out if there is something planned in terms of external-resource-watch and he said that there's a similar issue we're facing with the pull request when dealing with security group attachments.

I asked him if it would make sense to work on this and he said (quote):

```
i have been planning to have such controller to deal with security group attachment(currently under ip mode, the security group is detached immediately when pods is terminating, and breaks on-going connections)
if we have it, we can use it to deal with security group attachment and pod readiness gate together
```

not sure what that refers to; creating the controller on our own or waiting until the controller is there.

**Comment by nirnanaaa** [comment]
I will setup a new PR, to propose such a controller.

**Comment by billyshambrook** [comment]
Hey, we have ran into the issue described in this issue and have also tested how we should handle this when we are rotating our AMIs which require rotating all of our EC2 instances. We rely on cloudformation update policy to rotate our instances in an autoscaling group. This presents an additional factor that the ALB controller is getting rescheduled at the same time other pods are. We have currently found setting the prestop hook sleep to 60secs mostly avoids 5xx errors.

**Comment by arminc** [comment]
Me, @nirnanaaa and @iverberk were trying to implement something outside of the ALB (code is here: https://github.com/nirnanaaa/kube-readiness) to test the theories and this is what we found out:

1) You need to set `publishNotReadyAddresses` otherwise the pods are not registered to the ALB and they will never become ready.
2) When using the `PodReadinessGate` the pod status shows ready but it's not until you set the gate to true. When you want to know if the pod is ready you can't rely on the status alone you need to look at all statuses.
3) You need to put a terminating Pod to draining in ALB to make sure no NEW calls are sent to it before the Pod is gone.
4) To achieve number 3 your Pod needs to go into terminating state and stay in that state long enough for ALB to finish draining.
5) Most applications don't do number 4 properly so you end up adding a `preStop` hook.
6) Because of number 1 alb ingress controller will re-add the container after you put it to draining.
7) After fixing all of this we still notice that ALB is sending traffic to the pod (for at least 15 seconds) even while the pod is in draining state.

After doing all of this we were able to make it work eventually with consistent behaviour and achieving zero-downtime rolling update for our end-2-end test.

**Unsolved issues:**
1) Why does the ALB send traffic to draining pods? We think it might have to do something with keep-alive between the ALB and the pods but we don't know for sure until AWS explains what is happening. @M00nF1sh Did you already have contact with the ALB team about this?

**Conclusion:**
If the above issue is resolved it should be possible to tackle the problem

**Alternative:**
Use the current alb ingress controller and add a long enough `preStop` hook, mostly 30 seconds. Because the pod is removed from healthy targets in the endpoint the ingress controller will set the pod in draining in ALB. Because the pod is not removed directly (preStop hook) it will allow connections to drain.

**Comment by stevenpall** [comment]
@nirnanaaa et al.: were you able to get a new PR set up? It seems like this issue has stalled and it's a pretty large blocker for anyone wanting to use IP target mode in a production environment. Thanks!

**Comment by nirnanaaa** [comment]
@stevenpall unfortunately not. I am at this point uncertain where to continue. The repo, that @arminc mentioned is in theory working just fine. With the controller and a prestop hook it's working quite well. 
If someone has an idea I'd love to join with him.

**Comment by devkid** [comment]
Hi all,

I have updated my pull request: https://github.com/kubernetes-sigs/aws-alb-ingress-controller/pull/955 It now starts a go routine per target group to reconcile the condition status (it quits when all targets are healthy – see the code for more details).

Testing would be very welcome.
